File: apps/reading/src/api/writeModeTwinDraftMergeCompose.ts
/**
 * Write-mode twin draft merge compose (pure).
 *
 * Operator vision: recursive twin note-taker substrate of insights/questions
 * can be leveraged into writing — merge twin notes into a provisional HTML
 * write draft before full author commit.
 *
 * draft_written always false.
 * merge_executed always false.
 * store_mutated always false.
 */

export interface TwinWriteSlice {
  parent_asset_id: string;
  insights: string[];
  questions: string[];
}

export interface WriteModeTwinDraftMergeInput {
  draft_id: string;
  /** Optional existing write body (caller-supplied HTML/text). */
  base_draft_html?: string | null;
  slices: TwinWriteSlice[];
  operator_ack: boolean;
}

export interface WriteModeTwinDraftMergeCompose {
  draft_id: string;
  parent_asset_ids: string[];
  sections: string[];
  section_count: number;
  insight_count: number;
  question_count: number;
  /** True when ≥1 insight/question and operator_ack. */
  draft_ready: boolean;
  /** Always false — pure layer never writes draft assets. */
  draft_written: false;
  /** Always false — pure layer never merges into published write. */
  merge_executed: false;
  /** Always false — no store mutation. */
  store_mutated: false;
  notes: string[];
  authority: "write_mode_twin_draft_merge_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Build a provisional write draft from twin substrate slices.
 * Never invents insights/questions; never writes assets.
 */
export function composeWriteModeTwinDraftMerge(
  input: WriteModeTwinDraftMergeInput,
): WriteModeTwinDraftMergeCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const draft_id = requireNonEmpty(input.draft_id, "draft_id");
  if (!Array.isArray(input.slices) || input.slices.length === 0) {
    throw new Error("slices must be a non-empty array");
  }

  const notes: string[] = [
    "draft_written=false — provisional write draft not persisted",
    "merge_executed=false — published write not mutated",
    "store_mutated=false",
    "twin content is caller-supplied only (no invent)",
  ];

  const sections: string[] = [];
  if (input.base_draft_html != null && input.base_draft_html !== undefined) {
    if (
      typeof input.base_draft_html !== "string" ||
      !input.base_draft_html.trim()
    ) {
      throw new Error("base_draft_html must be non-empty string when set");
    }
    sections.push(
      `<section data-role="base-draft" data-draft="${draft_id}">${input.base_draft_html.trim()}</section>`,
    );
  } else {
    notes.push("base_draft_html absent — twin-only draft scaffold");
  }

  const parent_asset_ids: string[] = [];
  const seen = new Set<string>();
  let insight_count = 0;
  let question_count = 0;

  for (let i = 0; i < input.slices.length; i++) {
    const sl = input.slices[i];
    if (!sl || typeof sl !== "object") {
      throw new Error(`slices[${i}] must be an object`);
    }
    const parent = requireNonEmpty(
      sl.parent_asset_id,
      `slices[${i}].parent_asset_id`,
    );
    if (seen.has(parent)) {
      throw new Error(`duplicate parent_asset_id: ${parent}`);
    }
    seen.add(parent);
    parent_asset_ids.push(parent);

    if (!Array.isArray(sl.insights)) {
      throw new Error(`slices[${i}].insights must be an array`);
    }
    if (!Array.isArray(sl.questions)) {
      throw new Error(`slices[${i}].questions must be an array`);
    }
    for (let j = 0; j < sl.insights.length; j++) {
      const ins = requireNonEmpty(
        sl.insights[j],
        `slices[${i}].insights[${j}]`,
      );
      sections.push(
        `<section data-role="twin-insight" data-parent="${parent}">${ins}</section>`,
      );
      insight_count += 1;
    }
    for (let j = 0; j < sl.questions.length; j++) {
      const q = requireNonEmpty(
        sl.questions[j],
        `slices[${i}].questions[${j}]`,
      );
      sections.push(
        `<section data-role="twin-question" data-parent="${parent}">${q}</section>`,
      );
      question_count += 1;
    }
  }

  const hasTwin = insight_count + question_count >= 1;
  const draft_ready = input.operator_ack && hasTwin;

  notes.push(
    `parents=${parent_asset_ids.length} · insights=${insight_count} · questions=${question_count}`,
  );
  if (!input.operator_ack) {
    notes.push("draft_ready=false — operator_ack required");
  } else if (!hasTwin) {
    notes.push("draft_ready=false — no twin insights/questions (no invent)");
  } else {
    notes.push(
      `draft_ready=true · sections=${sections.length} (provisional only)`,
    );
  }

  notes.push("draft_written=false");
  notes.push("merge_executed=false");
  notes.push("store_mutated=false");

  return {
    draft_id,
    parent_asset_ids,
    sections,
    section_count: sections.length,
    insight_count,
    question_count,
    draft_ready,
    draft_written: false,
    merge_executed: false,
    store_mutated: false,
    notes,
    authority: "write_mode_twin_draft_merge_compose_advisory",
  };
}

export function formatWriteModeTwinDraftMergeSummary(
  c: WriteModeTwinDraftMergeCompose,
): string {
  return (
    `draft_ready=${c.draft_ready} · sections=${c.section_count} · ` +
    `insights=${c.insight_count} · questions=${c.question_count} · ` +
    `draft_written=false · merge_executed=false`
  );
}

File: substrate/write_mode_twin_draft_merge_compose.py
"""Write-mode twin draft merge compose (pure).

draft_written, merge_executed, store_mutated always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


class WriteModeTwinDraftMergeComposeError(ValueError):
    """Fail-closed validation for write-mode twin draft merge."""


@dataclass(frozen=True)
class WriteModeTwinDraftMergeCompose:
    draft_id: str
    parent_asset_ids: tuple[str, ...]
    sections: tuple[str, ...]
    section_count: int
    insight_count: int
    question_count: int
    draft_ready: bool
    draft_written: bool
    merge_executed: bool
    store_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "draft_id": self.draft_id,
            "parent_asset_ids": list(self.parent_asset_ids),
            "sections": list(self.sections),
            "section_count": self.section_count,
            "insight_count": self.insight_count,
            "question_count": self.question_count,
            "draft_ready": self.draft_ready,
            "draft_written": False,
            "merge_executed": False,
            "store_mutated": False,
            "notes": list(self.notes),
            "authority": "write_mode_twin_draft_merge_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise WriteModeTwinDraftMergeComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_write_mode_twin_draft_merge(
    *,
    draft_id: object,
    slices: object,
    operator_ack: object,
    base_draft_html: object | None = None,
) -> WriteModeTwinDraftMergeCompose:
    """Build provisional write draft from twin slices. Never writes assets."""
    if not isinstance(operator_ack, bool):
        raise WriteModeTwinDraftMergeComposeError(
            "operator_ack must be an explicit boolean"
        )
    did = _require_nonempty(draft_id, field="draft_id")
    if not isinstance(slices, list) or len(slices) == 0:
        raise WriteModeTwinDraftMergeComposeError(
            "slices must be a non-empty array"
        )

    notes: list[str] = [
        "draft_written=false — provisional write draft not persisted",
        "merge_executed=false — published write not mutated",
        "store_mutated=false",
        "twin content is caller-supplied only (no invent)",
    ]

    sections: list[str] = []
    if base_draft_html is not None:
        if not isinstance(base_draft_html, str) or not base_draft_html.strip():
            raise WriteModeTwinDraftMergeComposeError(
                "base_draft_html must be non-empty string when set"
            )
        sections.append(
            f'<section data-role="base-draft" data-draft="{did}">'
            f"{base_draft_html.strip()}</section>"
        )
    else:
        notes.append("base_draft_html absent — twin-only draft scaffold")

    parent_asset_ids: list[str] = []
    seen: set[str] = set()
    insight_count = 0
    question_count = 0

    for i, sl in enumerate(slices):
        if not isinstance(sl, dict):
            raise WriteModeTwinDraftMergeComposeError(
                f"slices[{i}] must be an object"
            )
        parent = _require_nonempty(
            sl.get("parent_asset_id"), field=f"slices[{i}].parent_asset_id"
        )
        if parent in seen:
            raise WriteModeTwinDraftMergeComposeError(
                f"duplicate parent_asset_id: {parent}"
            )
        seen.add(parent)
        parent_asset_ids.append(parent)
        insights = sl.get("insights")
        questions = sl.get("questions")
        if not isinstance(insights, list):
            raise WriteModeTwinDraftMergeComposeError(
                f"slices[{i}].insights must be an array"
            )
        if not isinstance(questions, list):
            raise WriteModeTwinDraftMergeComposeError(
                f"slices[{i}].questions must be an array"
            )
        for j, v in enumerate(insights):
            ins = _require_nonempty(v, field=f"slices[{i}].insights[{j}]")
            sections.append(
                f'<section data-role="twin-insight" data-parent="{parent}">'
                f"{ins}</section>"
            )
            insight_count += 1
        for j, v in enumerate(questions):
            q = _require_nonempty(v, field=f"slices[{i}].questions[{j}]")
            sections.append(
                f'<section data-role="twin-question" data-parent="{parent}">'
                f"{q}</section>"
            )
            question_count += 1

    has_twin = insight_count + question_count >= 1
    draft_ready = operator_ack and has_twin
    notes.append(
        f"parents={len(parent_asset_ids)} · insights={insight_count} · "
        f"questions={question_count}"
    )
    if not operator_ack:
        notes.append("draft_ready=false — operator_ack required")
    elif not has_twin:
        notes.append(
            "draft_ready=false — no twin insights/questions (no invent)"
        )
    else:
        notes.append(
            f"draft_ready=true · sections={len(sections)} (provisional only)"
        )
    notes.extend(
        ("draft_written=false", "merge_executed=false", "store_mutated=false")
    )

    return WriteModeTwinDraftMergeCompose(
        draft_id=did,
        parent_asset_ids=tuple(parent_asset_ids),
        sections=tuple(sections),
        section_count=len(sections),
        insight_count=insight_count,
        question_count=question_count,
        draft_ready=draft_ready,
        draft_written=False,
        merge_executed=False,
        store_mutated=False,
        notes=tuple(notes),
        authority="write_mode_twin_draft_merge_compose_advisory",
    )


__all__ = [
    "WriteModeTwinDraftMergeCompose",
    "WriteModeTwinDraftMergeComposeError",
    "compose_write_mode_twin_draft_merge",
]
